src/jit/morph.cpp

```diff
@@ -2656,55 +2608,32 @@ GenTree* Compiler::fgInsertCommaFormTemp(GenTree** ppTree, CORINFO_CLASS_HANDLE
 }
 
 //------------------------------------------------------------------------
-// fgMorphArgs: Walk and transform (morph) the arguments of a call
+// fgInitArgInfo: Construct the fgArgInfo for the call with the fgArgEntry for each arg
 //
 // Arguments:
-//    callNode - the call for which we are doing the argument morphing
+//    callNode - the call for which we are generating the fgArgInfo
 //
 // Return Value:
-//    Like most morph methods, this method returns the morphed node,
-//    though in this case there are currently no scenarios where the
-//    node itself is re-created.
+//    None
 //
 // Notes:
-//    This method is even less idempotent than most morph methods.
-//    That is, it makes changes that should not be redone. It uses the existence
-//    of gtCallLateArgs (the late arguments list) to determine if it has
-//    already done that work.
-//
-//    The first time it is called (i.e. during global morphing), this method
-//    computes the "late arguments". This is when it determines which arguments
-//    need to be evaluated to temps prior to the main argument setup, and which
-//    can be directly evaluated into the argument location. It also creates a
-//    second argument list (gtCallLateArgs) that does the final placement of the
-//    arguments, e.g. into registers or onto the stack.
-//
-//    The "non-late arguments", aka the gtCallArgs, are doing the in-order
-//    evaluation of the arguments that might have side-effects, such as embedded
-//    assignments, calls or possible throws. In these cases, it and earlier
-//    arguments must be evaluated to temps.
+//    This method is idempotent in that it checks whether the fgArgInfo has already been
+//    constructed, and just returns.
+//    This method only computes the arg table and arg entries for the call (the fgArgInfo),
+//    and makes no modification of the args themselves.
 //
-//    On targets with a fixed outgoing argument area (FEATURE_FIXED_OUT_ARGS),
-//    if we have any nested calls, we need to defer the copying of the argument
-//    into the fixed argument area until after the call. If the argument did not
-//    otherwise need to be computed into a temp, it is moved to gtCallLateArgs and
-//    replaced in the "early" arg list (gtCallArgs) with a placeholder node.
-
-#ifdef _PREFAST_
-#pragma warning(push)
-#pragma warning(disable : 21000) // Suppress PREFast warning about overly large function
-#endif
-GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
+void Compiler::fgInitArgInfo(GenTreeCall* call)
```

Comment:
This is truly a huge step forward. That being said I have a quick question. There was an expectation to call this several time? Upon the second call it would return as a no-op correct?

This would enable a call to `compiler::fgInitArgInfo` inside `compiler::canFastTailCall`

Reply:
Yes, that's the idea, though I confess that I wouldn't be at all surprised if getting that to work correctly required additional changes. Below, it checks whether the arg info has already been created (`if (call->fgArgInfo != nullptr)`), and simply returns.